### PR TITLE
WIP: Fix MINC read/write to convert from/to RAS coordinate system

### PR DESCRIPTION
**Pull request by gdevenyi**
Convert the coordinate system from RAS to LPS during MINC loading, and convert back on writing.

Also adds a setting to keep old configuration, so that MINC-based ITK tools (EZMinc) can maintain their original configuration and return sensible coordinate values in MINC-space.

Paging @vfonov

### CODE REVIEW

**Comment by gdevenyi** [comment]
Originally discussed at https://insightsoftwareconsortium.atlassian.net/browse/ITK-3503?attachmentSortBy=fileName

**Comment by vfonov** [comment]
What about minc transform?

**Comment by gdevenyi** [comment]
@vfonov I have the code for converting the MINC affine to from RAS to LPS, I was going to do a seperate PR.

**Comment by gdevenyi** [comment]
@vfonov MINC vs ITK also defines its transforms in opposite directions, do you want that added to the reader, or leave that to the user?

**Comment by gdevenyi** [comment]
I have added the conversion for Affine transforms.

I'm not quite sure where to add the switch to disable this conversion to support backwards compatibility.

Can anyone give some guidance here?

**Comment by gdevenyi** [comment]
My transform work isn't right. It used to be combined with inverting the affines so that MINC and ITK transforms went in the same direction. I backed that part out and this needs some fixes as a result.

**Comment by vfonov** [comment]
one of the main reason for me to have minc interface for ITK is to be able to run ANTs & Elastix. I think we should run extensive checks to make sure that this conversion does not screw up their results.

**Comment by gdevenyi** [comment]
Totally agree. On the walk home I wrote up a bunch of tests in head to ensure this works properly. Will enumerate later.

**Comment by gdevenyi** [comment]
Given two files, in both MINC and NIFTI format (converted using mnc2nii)

Tests to do:
- minc-to-minc affine registration with antsRegistration
- nifti-to-minc "
- minc-to-nifti "
- bestlinreg minc-to-minc
- save transforms as both ants and minc format

- apply all transforms using antsApplyTransforms
- apply minc-style transforms with mincresample

**Comment by vfonov** [comment]
Sounds like a plan

**Comment by gdevenyi** [comment]
Additional tests:
- antsRegistration SyN to produce nonlinear
- save as minc/nifti
- confirm forward/reverse transformations also work as expected with antsApplyTransforms and mincresample
- nlfit to produce nonlinear
- confirm forward/reverse work with antsApplyTransforms

**Comment by vfonov** [comment]
nlfit <-> antsApplyTransforms probably will not work.

**Comment by gdevenyi** [comment]
PR https://github.com/InsightSoftwareConsortium/ITK/pull/194 addresses the issue with  MINC NLIN transforms only being generated one-way and ITK not properly handling the invert flag on such transforms.

Still remaining is I need to figure out an efficient method to voxel-wise multiply the incoming grid file by (-1,-1,0) to flip the RAS vector components to LPS vectors.

File: Modules/IO/MINC/src/itkMINCImageIO.cxx [hunk 1]
```diff
@@ -456,6 +457,14 @@ void MINCImageIO::ReadImageInformation()
   dir_cos.Fill(0.0);
   dir_cos.SetIdentity();
 
+  //MINC stores direction cosines in RAS, need to convert to LPS for ITK
+  Matrix< double, 3,3 > RAS_tofrom_LPS;
+  RAS_tofrom_LPS.Fill(0.0);
```
Comment (dzenanz) [inline Modules/IO/MINC/src/itkMINCImageIO.cxx]: I think this operation is needless, as its result is annihilated by SetIdentity on the next line.

File: Modules/IO/TransformMINC/include/itkMINCTransformIO.hxx [hunk 2]
```diff
@@ -98,15 +98,45 @@ MINCTransformIOTemplate<TParametersValueType>
       ParametersType parameterArray;
       parameterArray.SetSize(12);
 
+
+      Matrix< double, 3,3 > RAS_affine_transform;
+      Matrix< double, 3,3 > LPS_affine_transform;
+      RAS_affine_transform.Fill(0.0);
```
Comment (dzenanz) [inline Modules/IO/TransformMINC/include/itkMINCTransformIO.hxx]: Probably not needed.

File: Modules/IO/TransformMINC/include/itkMINCTransformIO.hxx [hunk 3]
```diff
@@ -98,15 +98,45 @@ MINCTransformIOTemplate<TParametersValueType>
...
+      RAS_affine_transform.SetIdentity();
+      //MINC stores transforms in RAS, need to convert to LPS for ITK
+      Matrix< double, 3,3 > RAS_tofrom_LPS;
+      RAS_tofrom_LPS.Fill(0.0);
```
Comment (dzenanz) [inline Modules/IO/TransformMINC/include/itkMINCTransformIO.hxx]: Can be removed, I think.

File: Modules/IO/TransformMINC/include/itkMINCTransformIO.hxx [hunk 4]
```diff
@@ -231,14 +261,43 @@ MINCTransformIOTemplate<TParametersValueType>
       MatrixType matrix = matrixOffsetTransform->GetMatrix();
       OffsetType offset = matrixOffsetTransform->GetOffset();
 
+      //MINC stores everything in RAS, need to convert from LPS
+      Matrix< double, 3,3 > RAS_tofrom_LPS;
+      Matrix< double, 3,3 > RAS_affine_transform;
+      Matrix< double, 3,3 > LPS_affine_transform;
+      RAS_tofrom_LPS.Fill(0.0);
```
Comment (dzenanz) [inline Modules/IO/TransformMINC/include/itkMINCTransformIO.hxx]: Probably redundant.

**Comment by gdevenyi** [comment]
Thanks @dzenanz I have removed the extraneous Fills.

**Comment by gdevenyi** [comment]
@thewtex that's what I'm asking? how do I go about checking that locally :)

**Comment by thewtex** [comment]
@gdevenyi ah, turn on `BUILD_TESTING` in your ITK build's CMake configuration, re-build, then run `ctest` from the build tree.

**Comment by gdevenyi** [comment]
All the test are failing with errors like this, with can't read a file.
```
/scratch/ITK-build/bin/ITKIOMINCTestDriver "--compare" "/scratch/ITK-build/ExternalData/Modules/IO/MINC/test/Input/t1_z+_ushort_trans.mnc" "/scratch/ITK-build/Testing/Temporary/t1_z+_ushort_trans.mnc" "itkMINCImageIOTest4" "/scratch/ITK-build/ExternalData/Modules/IO/MINC/test/Input/t1_z+_ushort_trans.mnc" "/scratch/ITK-build/Testing/Temporary/t1_z+_ushort_trans.mnc" "427590.7957" "-8.195997123" "72.45943721" "-3.148635961"
Image:/scratch/ITK-build/ExternalData/Modules/IO/MINC/test/Input/t1_z+_ushort_trans.mnc sum=427590.7957 COM=[8.195997127, -72.45943721, -3.148635962]
Total mx mismatch:8.196
Image:/scratch/ITK-build/ExternalData/Modules/IO/MINC/test/Input/t1_z+_ushort_trans.mnc sum=427590.7957 COM=[8.195997127, -72.45943721, -3.148635962]
Total mx mismatch:8.196
/home/cic/devgab/projects/src/ITK.github/Modules/ThirdParty/MINC/src/libminc/libsrc2/volume.c:1400 (from MINC): Unable to open file '/scratch/ITK-build/Testing/Temporary/t1_z+_ushort_trans.mnc'
Exception detected while reading /scratch/ITK-build/ExternalData/Modules/IO/MINC/test/Input/t1_z+_ushort_trans.mnc : itk::ERROR: MINCImageIO(0x23e97a0): Could not open file "/scratch/ITK-build/Testing/Temporary/t1_z+_ushort_trans.mnc".<DartMeasurement name="BaselineImageName" type="text/string">INVALID_BASELINE_GIVEN</DartMeasurement>
```

It looks to me like the tests are getting setup slightly wrong maybe and the input files can't be found? Or am I misreading this?

**Comment by thewtex** [comment]
The test runs, then it tries to compare a test output file, */scratch/ITK-build/Testing/Temporary/t1_z+_ushort_trans.mnc* to a baseline file, but the test output file not exist. What is the test output file not being written?

**Comment by gdevenyi** [comment]
Oops. I see it now. Its testing COM in RAS space, and that's wrong now :+1: 

Patch incoming

**Comment by gdevenyi** [comment]
All passing!

@vfonov the only remaining question is:
do you need a flag to disable the correction of coordinate spaces for the MINCTransform code? I'm not sure how I would do that right now...

**Comment by vfonov** [comment]
yes, i would say that it is critical, and it have to be very explicit to **Enable** new behaviour

**Comment by vfonov** [comment]
> All passing!
> 
> @vfonov the only remaining question is:
> do you need a flag to disable the correction of coordinate spaces for the MINCTransform code? I'm not sure how I would do that right now...

Also,  i would like to completely disable the new behaviour where COM location is in different coordinates then what mincstats would report.

**Comment by gdevenyi** [comment]
@vfonov the disable coordinate conversion is already added at https://github.com/InsightSoftwareConsortium/ITK/pull/147/files#diff-32784fd54fb5bc9cf57e46a4395d785bR94

I have to *strongly* disagree that these fixes should be opt-in behaviour. Right now the ITK MINC reader is wrong. The coordinates are incorrect according to the internal ITK standard when loading files, making it impossible to mix MINC with any other filetype. These changes make MINC transparently work with the other filetypes ITKIO supports.

To provide backwards compatibility I will figure out how to add a flag to TransformMINC to disable this as well.

**Comment by vfonov** [comment]
If it was easy to disable, Wu did you have to change the tests that were failing?
I want to keep using old behavior by default.

**Comment by vfonov** [comment]
Well, of it will depend on me, I am not going to approve the code that breaks backward compatibility by default.
If ITK maintainers think that's the right thing to go, I will implement a patch for minc-toolkit to maintain compatibility.

**Comment by thewtex** [comment]
@vfonov @gdevenyi you have a better idea of what default makes sense regarding yielding correct behavior and preserving backwards compatibility. I can see arguments either way. Regardless, we can expose the option so the user can choose the desired behavior.

Thanks for the close attention and patches for this tricky issue.

**Comment by gdevenyi** [comment]
Additional changes coming:
Voxel-wise vector flipping should happen inside the MINCIO code, rather than the TransformMINC code. This is particularly motivated by the fact that I can pass a vector field directly into ANTs as a transform and if I gave it a MINC file with these patches it would do the wrong thing as the vector field wouldn't be flipped.

**Comment by hjmjohnson** [comment]
@gdevenyi  Is this worthy of finishing off?  It has been stale for 3 months now.

**Comment by gdevenyi** [comment]
Hi,

I got stuck at the point of how to handle flipping the individual loaded voxel coordinate systems right at load, rather than inside the TransformMINC code.

I think I need to do work after https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/IO/MINC/src/itkMINCImageIO.cxx#L134

But at this point we don't yet have an ITK image object, so I can't do ITK cleverness I did in the TransformMINC code.

**Comment by gdevenyi** [comment]
I am beyond my depth at this point to figure out how to flip the vector directions during load/write in the low-level IO code. Can anyone offer any advice?

**Comment by vfonov** [comment]
So, one can create a composite transform and then recalculate deformation field:
something like  
```
CompositeTransform->AddTransform(LPS_TO_RAS);
CompositeTransform->AddTransform(DisplacementFieldTransform);
CompositeTransform->AddTransform(RAS_TO_LPS);
...
out=TransformToDisplacementFieldFilter(CompositeTransform)
```

**Comment by gdevenyi** [comment]
Thanks vlad, I was already able to fix the directions on the displacement field, but my issue is, what about other 4D vector data like DTI? I was hoping to fix this at a low level where everything is uniform.

**Comment by vfonov** [comment]
well, that's highly application dependent. Obviously if vector represents RGB values, it should be passed as it is, and if it's spatial directions - then it will have to be modified. And if it is DTI acquisition - one will have to adjust gradient direction matrix.

**Comment by egarza** [comment]
Hi,
I also just had this problem. When my input was MINC the `antsMultivariateTemplateConstruction.sh` changed the orientation after finishing. The only solution is to convert to NIFTI before using the script. Isn't there a way for ANTs not to re-orient and just read MINC header?

Thanks

Eduardo

**Comment by vfonov** [comment]
Hello,

antsMultivariateTemplateConstruction.sh uses hard-coded .nii.gz file suffix. So, short answer - it is not meant to be used with .mnc files.

**Comment by egarza** [comment]
Right, I wasn't aware of this. 
But it did finish and I didn't get an error or warning. 

Just a thought, if it's not meant to use .mnc, perhaps the use of .mnc files should stop the script and give a warning or error?

**Comment by vfonov** [comment]
perhaps, here is the repository with that script: https://github.com/ANTsX/ANTs/blob/master/Scripts/antsMultivariateTemplateConstruction.sh - you can submit a pull-request

**Comment by hjmjohnson** [comment]
@gdevenyi Is this PR still relavant?

**Comment by gdevenyi** [comment]
> @gdevenyi Is this PR still relavant?

Yes, the MINC reader is still incorrect, and the coordinate system conversion fixes it.

**Comment by hjmjohnson** [comment]
@gdevenyi Could you provide a roadmap for what you think needs to be done still?  There may be students looking for projects this fall, and I'll see if I can convince one of them to look at this.

**Comment by gdevenyi** [comment]
@hjmjohnson

Current state as I recall, fixed are
- images load properly now
- affine transforms read and write properly (both MINC/ITK push/pull definitions handled and RAS/LPS coordinates converted
- warp transforms read and write properly (both MINC/ITK push/pull definitions handled and RAS/LPS coordinates converted

as I recall the blockers were
1. How to handle the breaking change for downstream dependent tools, right now there's a  bool option in the loader, but I don't see how to access given how most tools use a generic reader, I think it should be a compile-time switch instead, I'm not skilled enough in CMAKE to make that happen
2. How to handle other voxel formats which have directional information, in particular DTI

**Comment by vfonov** [comment]
There is  currently 4 different ways how Analyze file format can be handled, just make the same thing for MINC/XFM reader writer : https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/IO/NIFTI/include/itkNiftiImageIO.h#L42

**Comment by vfonov** [comment]
It  looks like the origin is incorrect when RAS to LPS is needed

**Comment by gdevenyi** [comment]
Good catch @vfonov I'll take a look

**Comment by vfonov** [comment]
I am trying to make a new pull on the latest codebase: https://github.com/vfonov/ITK/tree/MINC_RAS_LPS_v2 (WIP)

**Comment by gdevenyi** [comment]
> It looks like the origin is incorrect when RAS to LPS is needed

@vfonov The Origin Read/Write will need negative applied to the X and Y dimensions to go from R->L and A->P

**Comment by vfonov** [comment]
> > It looks like the origin is incorrect when RAS to LPS is needed
> 
> @vfonov The Origin Read/Write will need negative applied to the X and Y dimensions to go from R->L and A->P

So, most things are fixed in https://github.com/vfonov/ITK/tree/MINC_RAS_LPS_v2 , we need to make more tests there.
Also, the TransformationMINC for RAS->LPS is not quite complete .

**Comment by gdevenyi** [comment]
Can we perhaps revive this to integrate this breaking change for ITKv6?

**Comment by vfonov** [comment]
I already forgot what needs to be done. Do we want to, essentially make MINCs behave as if they were NIFTIs ?

**Comment by seanm** [comment]
>Can we perhaps revive this to integrate this breaking change for ITKv6?

@gdeveny Can you rebase on master and fix the conflicts?  I'd be interested to see how/if this breaks our app (Brainsight)...

**Comment by dzenanz** [comment]
Why not just document (via direction matrix) that the image is in RAS coordinate system?

**Comment by gdevenyi** [comment]
> Why not just document (via direction matrix) that the image is in RAS coordinate system?

Is this a suggestion to address this in code somehow? I'm not deeply familiar with ITK internals, so I'll need some hints.

My overall goal here was to bring MINC to first-class support in terms of it being treated equally and interchangeably with any other image loaded via ImageIO. Right now, I can't do this due to the misinterpreted coordinates.

**Comment by dzenanz** [comment]
I did not take a close look at the code. If you are shuffling array elements to "conform to LPS" standard, you don't need to do that. I see that you are fiddling with orientation matrices, so this PR might already be doing what I suggested.

**Comment by vfonov** [comment]
As far as I remember now the issue is with transformations, in particular with non-linear `_grid` files. That would need to be recomputed via F∘T∘F transformation on loading and saving any such file.

**Comment by gdevenyi** [comment]
> As far as I remember now the issue is with transformations, in particular with non-linear `_grid` files. That would need to be recomputed via F∘T∘F transformation on loading and saving any such file.

Correct, this was the last outstanding thing that needs fixing in this PR in a performant way.

**Comment by gdevenyi** [comment]
Rebasing was too difficult so I manually ported the changes into the new PR #4864

**Pull request closed without merging**